contracts/0.8.25/vaults/LazyOracle.sol

```diff
@@ -121,42 +173,151 @@ contract LazyOracle is ILazyOracle {
     function updateVaultData(
         address _vault,
         uint256 _totalValue,
-        int256 _inOutDelta,
         uint256 _feeSharesCharged,
         uint256 _liabilityShares,
         bytes32[] calldata _proof
     ) external {
         bytes32 leaf = keccak256(
-            bytes.concat(keccak256(abi.encode(_vault, _totalValue, _inOutDelta, _feeSharesCharged, _liabilityShares)))
+            bytes.concat(keccak256(abi.encode(_vault, _totalValue, _feeSharesCharged, _liabilityShares)))
         );
         if (!MerkleProof.verify(_proof, _storage().vaultsDataTreeRoot, leaf)) revert InvalidProof();
 
+        int256 inOutDelta;
+        (_totalValue, inOutDelta) = _handleSanityChecks(_vault, _totalValue, _liabilityShares);
+
         VaultHub(payable(LIDO_LOCATOR.vaultHub()))
             .applyVaultReport(
                 _vault,
                 _storage().vaultsDataTimestamp,
                 _totalValue,
-                _inOutDelta,
+                inOutDelta,
                 _feeSharesCharged,
                 _liabilityShares
             );
     }
 
+    /// @notice handle sanity checks for the vault lazy report data
+    /// @param _vault the address of the vault
+    /// @param _totalValue the total value of the vault in refSlot
+    /// @param _liabilityShares the liabilityShares of the vault in refSlot
+    /// @return the smoothed total value, inOutDelta in the refSlot
+    function _handleSanityChecks(address _vault, uint256 _totalValue, uint256 _liabilityShares) public returns (uint256, int256 _inOutDelta) {
+        VaultHub vaultHub = VaultHub(payable(LIDO_LOCATOR.vaultHub()));
+        VaultHub.VaultConnection memory connection = vaultHub.vaultConnection(_vault);
+        VaultHub.VaultRecord memory record = vaultHub.vaultRecord(_vault);
+
+        // 1. Calculate inOutDelta in the refSlot
+        int256 curInOutDelta = record.inOutDelta;
+        (uint256 refSlot, ) = CONSENSUS_CONTRACT.getCurrentFrame();
+        if (record.cachedRefSlot == refSlot) {
+            _inOutDelta = record.cachedInOutDelta;
+        } else {
+            _inOutDelta = curInOutDelta;
+        }
+
+        // 2. Sanity check for total value increase
+        _totalValue = _checkTotalValue(_vault, _totalValue, _inOutDelta, record);
+
+        // 3. Sanity check for dynamic total value underflow
+        if (int256(_totalValue) + curInOutDelta - _inOutDelta < 0) revert UnderflowInTotalValueCalculation();
+
+        // 4. Sanity check for liability shares increase
+        uint256 maxLiabilityShares = Math256.max(connection.shareLimit, record.liabilityShares);
+        if (_liabilityShares > maxLiabilityShares) revert LiabilitySharesExceedsLimit();
+
+        return (_totalValue, _inOutDelta);
+    }
+
+    /*
+        Here we need to introduce 2 concepts Safe and Unsafe fund:
+        1. Safe fund happens when the vault owner first tops up ETH balance of the vault contract (through fund() method) 
```

Comment:
probably highlight the difference:
1) inOutDelta is used
2) inOutDelta isn't changing